src/webrtc/peerconnection.spec.ts

```diff
@@ -81,22 +77,59 @@ describe('PeerConnection', function() {
     };
     createOfferSpy.and.returnValue(Promise.resolve(mockOffer));
 
-    // We mock setLocalDescription, because we want to check that it has not
-    // been called by the time the first signalling message is added to the
-    // handler queue.
-    var setLocalDescriptionSpy =
-      spyOn(mockRtcPeerConnection, 'setLocalDescription');
+    var setLocalDescriptionSpy = spyOn(mockRtcPeerConnection,
+        'setLocalDescription');
 
-    var pc = new peerconnection.PeerConnectionClass(
-        mockRtcPeerConnection, 'test');
+    var pc = new peerconnection.PeerConnectionClass(mockRtcPeerConnection,
+        'test');
     pc.negotiateConnection();
 
-    pc.signalForPeerQueue.setSyncNextHandler(
-        (message:signals.Message) => {
+    pc.signalForPeerQueue.setSyncNextHandler((message:signals.Message) => {
       expect(message.type).toEqual(signals.Type.OFFER);
       expect(mockRtcPeerConnection.setLocalDescription).not.toHaveBeenCalled();
       done();
     });
 
   });
+
+  it('do not set ICE candidates until setRemoteDescription has resolved', (done) => {
+    var createDataChannelSpy = spyOn(mockRtcPeerConnection, 'createDataChannel');
+    createDataChannelSpy.and.callFake((label:string, init:RTCDataChannelInit) => {
+      mockRtcPeerConnection.handleEvent('onnegotiationneeded');
+      return Promise.resolve('foo-channel-id');
+    });
+
+    var createOfferSpy = spyOn(mockRtcPeerConnection, 'createOffer');
+    var mockOffer :freedom_RTCPeerConnection.RTCSessionDescription = {
+      type: 'sdp',
+      sdp: 'mock:sdp'
+    };
+    createOfferSpy.and.returnValue(Promise.resolve(mockOffer));
+
+    spyOn(mockRtcPeerConnection, 'addIceCandidate').and.callThrough();
+
+    // Have setRemoteDescription() set a candidate *before it has resolved*.
+    var setRemoteDescriptionSpy = spyOn(mockRtcPeerConnection,
+        'setRemoteDescription').and.callFake(
+        (desc:freedom_RTCPeerConnection.RTCSessionDescription) => {
+      pc.handleSignalMessage({
+        type: signals.Type.CANDIDATE,
+        candidate: {
+          candidate: 'fakeCandidate'
+        }
+      });
+      expect(mockRtcPeerConnection.addIceCandidate).not.toHaveBeenCalled();
```

Comment:
As discussed, I added a check that `addIceCandidate` *is* called.

This required a bit of refactoring. PTAL!